src/libs/ajax/file-browser-providers/AzureBlobStorageFileBrowserProvider.ts

```diff
@@ -0,0 +1,185 @@
+}
+
+interface BlobListRequestOptions {
+  maxResults: number
+  marker?: string
+}
+
+const AzureBlobStorageFileBrowserProvider = ({ workspaceId, pageSize = 1000 }: AzureBlobStorageFileBrowserProviderParams): FileBrowserProvider => {
+  const storageDetailsPromise = AzureStorage().details(workspaceId)
+
+  const getNextPage = async <T>(params: AzureBlobStorageFileBrowserProviderGetPageParams<T>): Promise<IncrementalResponse<T>> => {
+    const {
+      isFirstPage,
+      tagName,
+      mapBlobOrBlobPrefix,
+      pageToken,
+      pendingItems = [],
+      prefix,
+      previousItems = [],
+      sasUrl,
+      signal,
+    } = params
+
+    let buffer = pendingItems
+    let nextPageToken = pageToken
+
+    if (nextPageToken || isFirstPage) {
+      do {
+        const requestOptions: BlobListRequestOptions = {
+          maxResults: pageSize
+        }
+        if (nextPageToken) {
+          requestOptions.marker = nextPageToken
+        }
+
+        const url = Utils.mergeQueryParams({
+          restype: 'container',
+          comp: 'list',
+          delimiter: '/',
+          prefix,
+          ...requestOptions,
+        }, sasUrl)
+
+        const response = await fetchOk(url, { signal })
+        const responseText = await response.text()
+        const responseXML = new window.DOMParser().parseFromString(responseText, 'text/xml')
+
+        const blobOrBlobPrefixElements = Array.from(responseXML.getElementsByTagName(tagName))
+        buffer = buffer.concat(blobOrBlobPrefixElements.map(mapBlobOrBlobPrefix))
+
+        nextPageToken = responseXML.getElementsByTagName('NextMarker').item(0)?.textContent || undefined
+      } while (buffer.length < pageSize && nextPageToken)
+    }
+
+    const items = previousItems.concat(buffer.slice(0, pageSize))
+    const nextPendingItems = buffer.slice(pageSize)
+    const hasNextPage = nextPendingItems.length > 0 || !!nextPageToken
+
+    return {
+      items,
+      getNextPage: hasNextPage ?
+        ({ signal } = {}) => getNextPage({
+          isFirstPage: false,
+          tagName,
+          mapBlobOrBlobPrefix,
+          pageToken: nextPageToken,
+          pendingItems: nextPendingItems,
+          prefix,
+          previousItems: items,
+          sasUrl,
+          signal
+        } as AzureBlobStorageFileBrowserProviderGetPageParams<T>) :
+        () => {
+          throw new Error('No next page')
+        },
+      hasNextPage
+    }
+  }
+
+  return {
+    getFilesInDirectory: async (path, { signal } = {}) => {
+      const { sas: { url: sasUrl } } = await storageDetailsPromise
+      return getNextPage({
+        isFirstPage: true,
+        tagName: 'Blob',
+        mapBlobOrBlobPrefix: blob => {
+          const name = blob.getElementsByTagName('Name').item(0)!.textContent!
+
+          const blobProperties = blob.getElementsByTagName('Properties').item(0)
+          const creationTime = blobProperties?.getElementsByTagName('Creation-Time').item(0)!.textContent!
```

Comment:
Do the exclamation marks mean that an error will be thrown if the variable is undefined?

Reply:
Yes. 

`.item(0)` [returns null if the collection is empty](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCollection/item#return_value) / if no elements with that tag were found. However, we assume that the [API response will include these tags](https://learn.microsoft.com/en-us/rest/api/storageservices/list-blobs?tabs=azure-ad), so the exclamation mark ([non-null assertion operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator)) changes the return type of `item(0)` from `Node | null` to `Node`. Similar with `textContent`: [Node.textContent](https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent) is typed `string | null`, so `textContent!` is typed `string`.

Just noticed that `blobProperties` has a `?` instead of a `!`. Will change that.